String memory fixes

R4PaSs opened the pull request:
Another step towards the closure of #1106 and #69, this PR removes several calls to `malloc` in `lib/standard` C code and replaces them by some of the solutions proposed in #1106 to handle the allocation of Strings.

@xymus: Your review is more than welcome since we're dealing with extern code, it might be a good time to specify and document how to properly allocate extern data structures to avoid leaking code such as the ones we had.

Morriar commented:
Ça donne quoi côté perfs?

privat commented on src/interpreter/naive_interpreter.nit:
```diff
@@ -857,8 +857,15 @@ redef class AMethPropdef
-				return v.native_string_instance(recvval.to_s)
+				var s = recvval.to_s
+				var srecv = args[1].val.as(Buffer)
+				for i in [0 .. s.length[ do
```
Ya pas de methode de copie?

R4PaSs replied:
Ouais, la que tu mets le doigt dessus, j'imagine un simple `clear` + `append` ferait la job. Je vais reroll.

xymus commented on lib/standard/string_nit.c:
```diff
@@ -11,10 +11,12 @@
 
+// Returns the length of `recv` as a `char*` (excluding the null character)
+long native_int_length_str(long recv){
+	return snprintf(NULL, 0, "%ld", recv);
```
Have you considered doing this in Nit? It would probably require using floats, but it would be interesting to compare the perf... or look at how snprintf is implemented.

~~~
var length = (self.pow10 + 1).floor
if self < 0 then length += 1
~~~

privat replied:
Ca pourrait marcher mais c'est fragile. snprintf est le mieux placé pour
mesurer la taille nécessaire du bluffer.
On pourrait imaginer des tas de saletés associés à la locale ou l'unicode
qui tomberaient comme ça d'un coup.

Une autre solution serait d'exposer `new NativeString` à la couche C qui
utiliserait ça au lieu du malloc pour allouer son buffer

On Wed, Feb 4, 2015 at 8:21 AM, Alexis Laferrière <notifications@github.com>
wrote:

> In lib/standard/string_nit.c
> <https://github.com/privat/nit/pull/1140#discussion_r24056200>:
>
> > @@ -11,10 +11,12 @@
> >
> >  #include "string_nit.h"
> >
> > +// Returns the length of `recv` as a `char*` (excluding the null character)
> > +long native_int_length_str(long recv){
> > +	return snprintf(NULL, 0, "%ld", recv);
>
> Have you considered doing this in Nit? It would probably require using
> floats, but it would be interesting to compare the perf... or look at how
> snprintf is implemented.
>
> var length = (self.pow10 + 1).floor
> if self < 0 then length += 1
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/privat/nit/pull/1140/files#r24056200>.
>

R4PaSs replied:
+1 pour l'exposition de new NativeString a la couche C, mais présentement
quand on fait ça dans un morceau de ffi, le compilo crashe parce que la
méthode existe pas.

Est-ce que je lève une issue pour ça ?
On 3 Feb 2015 20:37, "Jean Privat" <notifications@github.com> wrote:

> In lib/standard/string_nit.c
> <https://github.com/privat/nit/pull/1140#discussion_r24056821>:
>
> > @@ -11,10 +11,12 @@
> >
> >  #include "string_nit.h"
> >
> > +// Returns the length of `recv` as a `char*` (excluding the null character)
> > +long native_int_length_str(long recv){
> > +	return snprintf(NULL, 0, "%ld", recv);
>
> Ca pourrait marcher mais c'est fragile. snprintf est le mieux placé pour
> mesurer la taille nécessaire du bluffer.
> … <#14b523bcd11e6103_>
> On pourrait imaginer des tas de saletés associés à la locale ou l'unicode
> qui tomberaient comme ça d'un coup. Une autre solution serait d'exposer
> `new NativeString` à la couche C qui utiliserait ça au lieu du malloc pour
> allouer son buffer On Wed, Feb 4, 2015 at 8:21 AM, Alexis Laferrière <
> notifications@github.com> wrote: In lib/standard/string_nit.c <
> https://github.com/privat/nit/pull/1140#discussion_r24056200>: > @@
> -11,10 +11,12 @@ > > #include "string_nit.h" > > +// Returns the length of
> `recv` as a `char*` (excluding the null character) > +long
> native_int_length_str(long recv){ > + return snprintf(NULL, 0, "%ld",
> recv); Have you considered doing this in Nit? It would probably require
> using floats, but it would be interesting to compare the perf... or look at
> how snprintf is implemented. var length = (self.pow10 + 1).floor if self <
> 0 then length += 1 — Reply to this email directly or view it on GitHub <
> https://github.com/privat/nit/pull/1140/files#r24056200>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/privat/nit/pull/1140/files#r24056821>.
>

privat replied:
> +1 pour l'exposition de new NativeString a la couche C, mais présentement quand on fait ça dans un morceau de ffi, le compilo crashe parce que la méthode existe pas. Est-ce que je lève une issue pour ça ?

Quand le compilateur crashe, ça devrait être issue sytématique

R4PaSs replied:
OK, je fais ça tantôt alors !
On 3 Feb 2015 20:44, "Jean Privat" <notifications@github.com> wrote:

> In lib/standard/string_nit.c
> <https://github.com/privat/nit/pull/1140#discussion_r24057161>:
>
> > @@ -11,10 +11,12 @@
> >
> >  #include "string_nit.h"
> >
> > +// Returns the length of `recv` as a `char*` (excluding the null character)
> > +long native_int_length_str(long recv){
> > +	return snprintf(NULL, 0, "%ld", recv);
>
>  +1 pour l'exposition de new NativeString a la couche C, mais présentement
> quand on fait ça dans un morceau de ffi, le compilo crashe parce que la
> méthode existe pas. Est-ce que je lève une issue pour ça ?
> Quand le compilateur crashe, ça devrait être issue sytématique
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/privat/nit/pull/1140/files#r24057161>.
>

R4PaSs commented:
Rerolled, replaced the copy loop by an append since no copy_to method is available between String and Buffer and FlatString.items is not available.

Also fixed a bug in copy_to intern call for NativeString, as it reported incorrectly invalid accesses.

R4PaSs commented:
@Morriar : I just checked on my machine, the impact on performance seems negligible.
Memory use on the other hand falls from 200M+ to 1M+ for the example program in #69 !

Morriar commented:
+1 in this case.

xymus commented:
+1
